README.md
# Checkpoint
## How to use
Before compiling or execute, create the file /src/config/config.js and include your credentials like the exemple below.

```javascript
module.exports = {
    "url": "",
    "company": '1',
    "user": "",
    "pass": ""
}
```

CHANGELOG.md
# CHANGELOG

## 2.0.0

- Structure upgraded to work with React
- No need to hardcode user and password

## 1.0.0

- MVP


app/main.dev.ts
/* eslint global-require: off, no-console: off */

/**
 * This module executes inside of electron's main process. You can start
 * electron renderer process from here and communicate with the other processes
 * through IPC.
 *
 * When running `yarn build` or `yarn build-main`, this file is compiled to
 * `./app/main.prod.js` using webpack. This gives us some performance wins.
 */
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import path from 'path';
import { app, BrowserWindow, screen, ipcMain } from 'electron';
import { autoUpdater } from 'electron-updater';
import log from 'electron-log';
import MenuBuilder from './menu';

import Norber from './utils/Norber'

export default class AppUpdater {
  constructor () {
    log.transports.file.level = 'info';
    autoUpdater.logger = log;
    autoUpdater.checkForUpdatesAndNotify();
  }
}

let mainWindow: BrowserWindow | null = null;

if (process.env.NODE_ENV === 'production') {
  const sourceMapSupport = require('source-map-support');
  sourceMapSupport.install();
}

if (
  process.env.NODE_ENV === 'development' ||
  process.env.DEBUG_PROD === 'true'
) {
  require('electron-debug')();
}

const installExtensions = async () => {
  const installer = require('electron-devtools-installer');
  const forceDownload = !!process.env.UPGRADE_EXTENSIONS;
  const extensions = ['REACT_DEVELOPER_TOOLS', 'REDUX_DEVTOOLS'];

  return Promise.all(
    extensions.map((name) => installer.default(installer[name], forceDownload))
  ).catch(console.log);
};

const createWindow = async () => {
  if (
    process.env.NODE_ENV === 'development' ||
    process.env.DEBUG_PROD === 'true'
  ) {
    await installExtensions();
  }

  mainWindow = new BrowserWindow({
    show: false,
    width: 260,
    height: 320,
    webPreferences:
      (process.env.NODE_ENV === 'development' ||
        process.env.E2E_BUILD === 'true') &&
        process.env.ERB_SECURE !== 'true'
        ? {
          nodeIntegration: true,
        }
        : {
          preload: path.join(__dirname, 'dist/renderer.prod.js'),
        },
  });

  const { x, y, width, height } = mainWindow.getBounds()

  mainWindow.setBounds({
    x: screen.getPrimaryDisplay().bounds.width - (width + 20),
    y: screen.getPrimaryDisplay().bounds.height - (height + 50),
    height,
    width
  })

  mainWindow.loadURL(`file://${__dirname}/app.html`);

  // @TODO: Use 'ready-to-show' event
  //        https://github.com/electron/electron/blob/master/docs/api/browser-window.md#using-ready-to-show-event
  mainWindow.webContents.on('did-finish-load', () => {
    if (!mainWindow) {
      throw new Error('"mainWindow" is not defined');
    }
    if (process.env.START_MINIMIZED) {
      mainWindow.minimize();
    } else {
      // norber.getCheckpoints().then((checkpoint) => {
      //   console.log("checkpoint", checkpoint)
      //   mainWindow.webContents.send('checkpoint:result', checkpoint)
      // })
      mainWindow.show();
      mainWindow.focus();
    }
  });

  mainWindow.on('closed', () => {
    mainWindow = null;
  });

  const menuBuilder = new MenuBuilder(mainWindow);
  menuBuilder.buildMenu();

  // Remove this if your app does not use auto updates
  // eslint-disable-next-line
  new AppUpdater();
};

/**
 * Add event listeners...
 */

app.on('window-all-closed', () => {
  // Respect the OSX convention of having the application in memory even
  // after all windows have been closed
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

if (process.env.E2E_BUILD === 'true') {
  // eslint-disable-next-line promise/catch-or-return
  app.whenReady().then(createWindow);
} else {
  app.on('ready', createWindow);
}

app.on('activate', () => {
  // On macOS it's common to re-create a window in the app when the
  // dock icon is clicked and there are no other windows open.
  if (mainWindow === null) createWindow();
});

app.allowRendererProcessReuse = false

ipcMain.on('checkpoint:get', (event, arg) => {
  console.log('checkpoint get')


  const norber = new Norber(arg.user, arg.password, arg.company, arg.url)

  norber.getCheckpoints().then((checkpoint) => {
    mainWindow.webContents.send('checkpoint:getReturn', checkpoint)
  })

})
ipcMain.on('checkpoint:set', (event, arg) => {
  console.log('checkpoint set')

  const norber = new Norber(arg.user, arg.password, arg.company, arg.url)

  norber.checkpoint().then((result) => {
    // event.sender.send('checkpoint:getReturnFromCheckpoint', result)
    norber.getCheckpoints().then((checkpoint) => {
      mainWindow.webContents.send('checkpoint:getReturn', checkpoint)
    })
  })

})

src/helpers/norber.js
const config = require('../config/config.js')

const puppeteer = require('puppeteer')

module.exports = class Norber {
    constructor() {
        this.browser;
        this.page;
    }

    today() {
        const date = new Date()
        const day = date.getDate() < 10 ? `0${date.getDate()}` : date.getDate()
        const month = (date.getMonth() + 1) < 10 ? `0${date.getMonth() + 1}` : (date.getMonth() + 1)
        const year = date.getFullYear() < 10 ? `0${date.getFullYear()}` : date.getFullYear()
        return `${day}/${month}/${year}`
    }

    async extractLine(content, line_number) {
        let cleanContent = content.replace(/ {2,}/g, ' ')
        const line = cleanContent.split('\n')
        console.log(line[line_number])
        const data = line[line_number].split(' ')
        const weekday = data[0]
        const date = data[1]
        const entrada = data[5].includes(':') ? data[5] : "00:00"
        const almoco = data[6].includes(':') ? data[6] : "00:00"
        const retorno = data[7].includes(':') ? data[7] : "00:00"
        const saida = data[8].includes(':') ? data[8] : "00:00"

        return { date, entrada, almoco, retorno, saida }
    }
    async login() {
        this.browser = await puppeteer.launch({ headless: true, executablePath: 'C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe' })
        this.page = await this.browser.newPage()
        await this.page.goto(config.url)
        const html = await this.page.content()

        await this.page.select('select[name=requiredempresa]', config.company)
        await this.page.type('input[name=requiredusuario]', config.user, { delay: 20 })
        await this.page.type('input[name=requiredsenha]', config.pass, { delay: 20 })
        await this.page.click('body > form > div.quadroLogin_Borda > div.quadroLogin_Conteudo > div:nth-child(3) > div.quadroLogin_Campo > input.BotaoAchatado')
    }

    async getCheckpoints() {
        console.log("Getting Checkpoints")
        console.log("Log in")
        await this.login()
        await this.page.waitForSelector("a#menu4.menuBotao")
        console.log("Logged")
        await this.page.click('a#menu4.menuBotao')
        await this.page.click('#menu4_Item2 a')

        let newPagePromise;

        newPagePromise = new Promise(x => this.browser.once('targetcreated', target => x(target.page())));
        const popup = await newPagePromise;
        await popup.waitForSelector('#rbOpSaida1')
        await popup.waitForSelector('#dtInicio')
        await popup.waitForSelector('#dtFim')
        await popup.waitForSelector('#Submit')

        await popup.click('#rbOpSaida1')
        const getToday = this.today()
        await popup.$eval('#dtInicio', (el, today) => el.value = today, getToday)
        await popup.$eval('#dtFim', (el, today) => el.value = today, getToday)
        const newPagePromise2 = new Promise(x => this.browser.once('targetcreated', target => x(target.page())));
        await popup.click('#Submit')
        const popup2 = await newPagePromise2

        const allPages = await this.browser.pages()
        console.log(allPages.length)

        await popup2.waitFor('pre')
        const content = await popup2.$eval('pre', element => element.innerHTML)
        console.log("content", content)
        await this.browser.close()

        const extractedData = await this.extractLine(content, 17)
        console.log(extractedData)
        return extractedData

    }
    async checkpoint() {
        console.log("Making Checkpoint")
        console.log("Login")
        await this.login()
        await this.page.waitForSelector("a#menu2.menuBotao")
        console.log("Logged")
        await this.page.click('a#menu2.menuBotao')

        const newPagePromise = new Promise(x => this.browser.once('targetcreated', target => x(target.page())));
        await this.page.click('#menu2_Item1 a')

        const popup = await newPagePromise;
        await popup.waitFor(2000)

        const values = await popup.evaluate(() => {
            let data = document.getElementById("data").value;
            let hour = document.querySelector("input[name=hora]").value

            return { data, hour }
        })
        popup.click('#Button1')
        await popup.waitFor(2000)

        await this.browser.close()
        return values
    }
}


src/assets/js/scripts.js
const { ipcRenderer } = require("electron");

const form_checkpoing = document.getElementById('form_checkpoint')
const btn_checkpoint = document.getElementById('btn_checkpoint')
const div_hour = document.getElementById('hour')

const td_entrada = document.getElementById('td_entrada')
const td_almoco = document.getElementById('td_almoco')
const td_retorno = document.getElementById('td_retorno')
const td_saida = document.getElementById('td_saida')
const loading = document.getElementById('loading')

// let entrada = "00:00";
// let almoco = "00:00";
// let retorno = "00:00";
// let saida = "00:00";

ipcRenderer.on('checkpoint:result', (event, arg) => {
    console.log("IPCRender")
    entrada = arg.entrada
    almoco = arg.almoco
    retorno = arg.retorno
    saida = arg.saida
    console.log(entrada, almoco, retorno, saida)

    fillHours(entrada, almoco, retorno, saida)
})

ipcRenderer.on('checkpoint:return', (event, arg) => {
    console.log(event, arg)
    loading.style.display = "none";
    if (entrada == "00:00") {
        entrada = arg.hour;
        td_entrada.innerText = entrada
        return
    } else if (almoco == "00:00") {
        almoco = arg.hour
        td_almoco.innerHTML = almoco
        return
    } else if (retorno == "00:00") {
        retorno = arg.hour
        td_retorno.innerHTML = retorno
        return
    } else {
        saida = arg.hour
        td_saida.innerHTML = saida
        return
    }
})

form_checkpoint.addEventListener('submit', (event) => {
    event.preventDefault();
    ipcRenderer.send('checkpoint:add')
    loading.style.display = "block";
})

const time = () => {
    const now = new Date()
    console.log("time")
    const hours = now.getHours() < 10 ? "0" + now.getHours() : now.getHours();
    const minutes = now.getMinutes() < 10 ? "0" + now.getMinutes() : now.getMinutes();
    const seconds = now.getSeconds() < 10 ? "0" + now.getSeconds() : now.getSeconds();

    div_hour.innerHTML = `${hours}:${minutes}:${seconds}`
}

const formatHour = (hour) => {
    let h = hour.getHours() < 10 ? '0' + hour.getHours() : hour.getHours()
    let m = hour.getMinutes() < 10 ? '0' + hour.getMinutes() : hour.getMinutes()
    return `${h}:${m}`;
}

const fillHours = (entrada, almoco, retorno, saida) => {
    const e = new Date(new Date().toDateString() + ' ' + entrada)
    const a = new Date(new Date().toDateString() + ' ' + almoco)
    const r = new Date(new Date().toDateString() + ' ' + retorno)
    const s = new Date(new Date().toDateString() + ' ' + saida)

    td_entrada.innerText = entrada
    td_almoco.innerText = almoco
    td_retorno.innerText = retorno
    td_saida.innerText = saida

    const p8h = 8 * 1000 * 60 * 60
    const p1h = 1 * 1000 * 60 * 60

    if (entrada !== "00:00" && almoco == "00:00" && retorno == "00:00") {
        const p1 = formatHour(new Date(e.getTime() + p8h + p1h))
        td_saida.innerText = 'P ' + p1
    } else if (entrada !== "00:00" && almoco !== "00:00" && retorno == "00:00") {
        const p1 = formatHour(new Date(e.getTime() + p8h + p1h))
        td_saida.innerText = 'P ' + p1
    } else if (entrada !== "00:00" && almoco !== "00:00" && retorno !== "00:00") {
        const tI = r.getTime() - a.getTime()
        const p1 = formatHour(new Date(new Date(e.getTime() + p8h + tI)))
        td_saida.innerText = 'P ' + p1
    }
}

const main = () => {
    setInterval(time, 1000);
    fillHours("00:00", "00:00", "00:00", "00:00")
}

main()

index.js
const { app, BrowserWindow, ipcMain, Tray, Menu, screen } = require('electron')
const path = require('path')
const Norber = require('./src/helpers/norber')

let mainWindow;

function createWindow() {
    mainWindow = new BrowserWindow({
        width: 260,
        height: 320,
        frame: false,
        show: true,
        alwaysOnTop: true,
        darkTheme: true,
        skipTaskbar: true,
        webPreferences: {
            nodeIntegration: true
        }
    })
    const { x, y, height, width } = mainWindow.getBounds()

    mainWindow.setBounds({
        x: screen.getPrimaryDisplay().bounds.width - (width + 20),
        y: screen.getPrimaryDisplay().bounds.height - (height + 50),
        height,
        width
    })

    mainWindow.loadFile(path.join(__dirname, 'src', 'index.html'))
    mainWindow.webContents.openDevTools({ detached: true })

}

app.on('ready', () => {
    createWindow();
    const norber = new Norber()
    norber.getCheckpoints().then((checkpoint) => {
        console.log("checkpoint", checkpoint)
        mainWindow.webContents.send('checkpoint:result', checkpoint)
    })

    if (app.dock) app.dock.hide()

    const contextMenu = Menu.buildFromTemplate([
        {
            label: 'Quit',
            acelerator: 'Ctrl+Q',
            click: () => { app.quit() }
        }
    ])

    tray = new Tray(path.join(__dirname, 'src', 'assets', 'img', 'clock.png'));
    tray.setToolTip('Checkpoint');
    tray.on('click', () => {
        if (mainWindow.isVisible()) {
            mainWindow.hide()
        } else {
            mainWindow.show()
        }
    })
    tray.setContextMenu(contextMenu)
})
app.on('window-all-closed', () => {
    if (process.platform !== 'darwin') {
        app.quit()
    }
})
app.on('activate', () => {
    if (BrowserWindow.getAllWindows().length === 0) {
        createWindow()
    }
})
ipcMain.on('checkpoint:add', (event, arg) => {
    console.log('recebido')
    const norber = new Norber()

    norber.checkpoint().then((result) => {
        event.sender.send('checkpoint:return', result)

        norber.getCheckpoints().then((checkpoint) => {
            console.log("checkpoint", checkpoint)
            mainWindow.webContents.send('checkpoint:result', checkpoint)
        })
    })


})

setInterval(() => {
    console.log("interval")
    const norber = new Norber()
    norber.getCheckpoints().then((checkpoint) => {
        console.log("checkpoint", checkpoint)
        mainWindow.webContents.send('checkpoint:result', checkpoint)
    })
}, 20 * 60 * 1000)